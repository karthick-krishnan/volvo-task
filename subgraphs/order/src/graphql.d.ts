declare module '*/order.graphql' {
    import { DocumentNode } from 'graphql';
    const MyQuery: DocumentNode;
  
    export { MyQuery };
  
    export default defaultDocument;
  }