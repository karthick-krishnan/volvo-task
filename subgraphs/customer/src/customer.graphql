type Customer {
  customerId: ID!
  name: String
  emailId: String
  phone: String
  address: String
}

extend type Query {
  getAllCustomers: [Customer]
}