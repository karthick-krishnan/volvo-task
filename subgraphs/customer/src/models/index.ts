'use strict';

import { Sequelize } from 'sequelize';
const env = process.env.NODE_ENV || 'development';
import config from '../config/config'


interface DBObject {
  [key: string]: any
}
const db: DBObject = {};


export const sequelize: Sequelize  = new Sequelize(config[env].database, config[env].username, config[env].password, config[env]);

db.sequelize = sequelize;

export default db;