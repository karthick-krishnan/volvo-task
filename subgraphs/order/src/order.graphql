type Customer {
  customerId: ID!
  name: String
  emailId: String
  phone: String
  address: String
}

type Product {
  prodId: ID!
  name: String
  qty: Int
  price: Float
}


type Order @key(fields: "orderId") {
  orderId: ID!
  customer: Customer
  products: [Product]
  deliveryDate: String
  totalPrice: Float
}

extend type Query {
  order(orderId: ID!): Order
}

extend type Mutation {
   order(customerId: ID, products: [ID], deliveryDate: String, totalPrice: Float, qty: Int): Order
}