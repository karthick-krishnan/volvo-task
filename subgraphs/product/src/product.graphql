type Product {
  prodId: ID!
  name: String
  qty: Int
  price: Float
}

extend type Query {
  allProducts: [Product]
  product(prodId: ID!): Product
}

extend type Mutation {
  createProduct(name: String, qty: Int, price: Float): Product
}