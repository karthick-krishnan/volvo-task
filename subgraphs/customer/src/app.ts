import { ApolloServer, gql } from "apollo-server";
import { buildFederatedSchema } from "@apollo/federation";
import { Customer } from "../src/models/customer";

const port = process.env.APOLLO_PORT || 4000;

const typeDefs = gql(
  `
  type Customer {
  customerId: ID!
  name: String
  emailId: String
  phone: String
  address: String
}

extend type Query {
  getAllCustomers: [Customer]
}
`
);

const resolvers = {
  Query: {
      getAllCustomers: async () => {
          return await Customer.findAll();
      }
  },
}


const server = new ApolloServer({ schema: buildFederatedSchema({ typeDefs, resolvers }) });


server
  .listen({ port: port })
  .then((type: any) => {
    console.log(`ðŸš€ Customer subgraph ready at ${type.url}`);
  })
  .catch((err: any) => {
    console.error(err);
  });
