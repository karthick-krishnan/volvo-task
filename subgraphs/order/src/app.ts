import { ApolloServer, gql } from "apollo-server";
import { buildFederatedSchema } from "@apollo/federation";
import {OrderResolvers} from './services/resolvers/order'

const port = process.env.APOLLO_PORT || 4000;


const type = `
type Customer {
  customerId: ID!
  name: String
  emailId: String
  phone: String
  address: String
}

type Product {
  prodId: ID!
  name: String
  qty: Int
  price: Float
}


type Order @key(fields: "orderId") {
  orderId: ID!
  customer: Customer
  products: [Product]
  deliveryDate: String
  totalPrice: Float
}

extend type Query {
  order(orderId: ID!): Order
}

extend type Mutation {
  order(customerId: ID, products: [ID], deliveryDate: String, totalPrice: Float, qty: Int): Order
}`;

export const typeDefs = gql`${type}`;

const server = new ApolloServer({
  schema: buildFederatedSchema({ typeDefs, resolvers : OrderResolvers }),
});


server
  .listen({ port: port })
  .then(({ url }: { url: string }) => {
    console.log(`ðŸš€ Order subgraph ready at ${url}`);
  })
  .catch((err) => {
    console.error(err);
  });
