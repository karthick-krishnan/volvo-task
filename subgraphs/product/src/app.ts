import { ApolloServer, gql } from "apollo-server";
import { buildFederatedSchema } from "@apollo/federation";
import { GraphQLError } from "graphql";
import { ProductResolvers } from "./services/resolvers/product";
const port = process.env.APOLLO_PORT || 4000;
const { readFileSync } = require("fs");
import LoggerPlugin from "./utils/logger-plugin";

// this can also be a glob pattern to match multiple files!

const type = `
type Product {
    prodId: ID!
    name: String
    qty: Int
    price: Float
  }
  
  extend type Query {
    allProducts: [Product]
    product(prodId: ID!): Product
  }
  
  extend type Mutation {
    createProduct(name: String, qty: Int, price: Float): Product
  }`;

export const typeDefs = gql(type);

export const server = new ApolloServer({
  schema: buildFederatedSchema({ typeDefs, resolvers: ProductResolvers }),
  plugins: [LoggerPlugin],
});

server
  .listen({ port: port })
  .then(({ url }) => {
    console.log(`ðŸš€ product subgraph ready at ${url}`);
  })
  .catch((err) => {
    console.error(err);
  });
