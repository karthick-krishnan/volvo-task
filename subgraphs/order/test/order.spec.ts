import { ApolloServer } from 'apollo-server';
import { buildFederatedSchema } from '@apollo/federation';
import { typeDefs } from '../src/app';
import { mockResolver, customerId, productId, noExistingOrderId } from "./mock";

const getOrderQuery = `query Order($orderId: ID!) {
  order(orderId: $orderId) {
    products {
      name
      price
    },
    customer {
      name,
      phone
      emailId
    }
    deliveryDate
    orderId
    totalPrice
  }
}
`

const createOrderMutation = `mutation Order($customerId: ID, $deliveryDate: String, $products: [ID], $qty: Int) {
  order(customerId: $customerId, deliveryDate: $deliveryDate, products: $products, qty: $qty) {
    customer {
      name
      emailId
    }
    products {
      name
      price
    }
    orderId
    totalPrice
  }
}
`
const noExistingCustomerId = "cd43965a-7ed2-11ec-90d6-0242ac120003";
let server: ApolloServer;

beforeAll(() => {
  console.log('Started')
  server = new ApolloServer({
    schema: buildFederatedSchema({ typeDefs, resolvers: mockResolver }),
  });
});

afterAll((done) => {
  server.stop();
  done();
});

describe("Order subgraph", () => {
  it("Gets the order", async () => {
    const res = await server.executeOperation({ query: getOrderQuery, variables: {orderId: "69005202-4dcf-4169-86f4-69a8e0aeb6ff"} });
    expect(res).toMatchSnapshot();
  });

  it("Gets the correct product details", async () => {
    const res = await server.executeOperation({ query: getOrderQuery, variables: {orderId: "69005202-4dcf-4169-86f4-69a8e0aeb6ff"} });
    expect(res?.data?.order?.totalPrice).toBe(25);
    expect(res?.data?.order?.products?.length).toBe(2);
    expect(res?.data?.order?.products[0]?.name).toBe("Type C Cable");
    expect(res?.data?.order?.products[1]?.name).toBe("Type C Charger");
  });

  it('Gets the correct customer details', async () => {
    const res = await server.executeOperation({ query: getOrderQuery, variables: {orderId: "69005202-4dcf-4169-86f4-69a8e0aeb6ff"} });
    expect(res?.data?.order?.customer?.emailId).toBe("karthick.taker@gmail.com");
    expect(res?.data?.order?.customer?.name).toBe("Karthick");
  })
  
  it("order does not exisit", async () => {
    const res: any = await server.executeOperation({ query: getOrderQuery, variables: {orderId: noExistingOrderId} });
    expect(res?.errors[0]?.message).toBe("orderId does not exist!!!!");
    expect(res?.data?.order).toBe(null);
  })

  it("Create a new order", async () => {
    const res = await server.executeOperation({ query: createOrderMutation, variables: {customerId , deliveryDate: "2022-02-18", products: [productId], qty: 1} });
    expect(res).toMatchSnapshot();
  });

  it("Create order throws out of stock error", async () => {
    const res: any = await server.executeOperation({ query: createOrderMutation, variables: {customerId , deliveryDate: "2022-02-18", products: [productId], qty: 26} });
    expect(res?.errors[0]?.message).toBe("product quantity quoted should be less!!!")
  });
});
