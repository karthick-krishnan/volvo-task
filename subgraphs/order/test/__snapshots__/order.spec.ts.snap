// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Order subgraph Create a new order 1`] = `
Object {
  "data": Object {
    "order": Object {
      "customer": Object {
        "emailId": "karthick.taker@gmail.com",
        "name": "Karthick",
      },
      "orderId": "69005202-4dcf-4169-86f4-69a8e0aeb6ff",
      "products": Array [
        Object {
          "name": "Type C Cable",
          "price": 12,
        },
      ],
      "totalPrice": 12,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Order subgraph Gets the order 1`] = `
Object {
  "data": Object {
    "order": Object {
      "customer": Object {
        "emailId": "karthick.taker@gmail.com",
        "name": "Karthick",
        "phone": "1234234",
      },
      "deliveryDate": "1643120622000",
      "orderId": "69005202-4dcf-4169-86f4-69a8e0aeb6ff",
      "products": Array [
        Object {
          "name": "Type C Cable",
          "price": 12,
        },
        Object {
          "name": "Type C Charger",
          "price": 13,
        },
      ],
      "totalPrice": 25,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
